src/controllers/auth.controllers.js

```diff
@@ -0,0 +1,128 @@
+import { validationResult } from 'express-validator';
+import bcrypt from 'bcryptjs';
+import { generateToken } from '../utils/authentication';
+import User from '../models/user.model';
+import Error from '../middleware/error/ErrorHandler';
+
+export const signup = async (req, res, next) => {
+  try {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+      next(
+        Error.unprocessableEntity('Please provide a valid username or password'),
+      );
+      return;
+    }
+    const userExist = await User.findOne({ email: req.body.email });
+    if (userExist) {
+      next(
+        Error.badRequest('Email already exists'),
+      );
+      return;
+    }
+    const salt = await bcrypt.genSalt();
+    const hashPassword = await bcrypt.hash(req.body.password, salt);
+    const user = new User({
+      username: req.body.username,
+      email: req.body.email,
+      password: hashPassword,
+    });
+    const { username, email, password } = req.body;
+    if (!username || !email || !password) {
+      next(
+        Error.badRequest('All fields are required'),
+      );
+      return;
+    }
+    await user.save();
+    // create and assign a token
+    const token = generateToken(user);
+    res.send({
+      success: true,
+      user,
+      token,
+    });
+  } catch (e) {
+    res.status(400).send({ success: false, e });
+  }
+};
+
+export const login = async (req, res, next) => {
+  try {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+      next(
+        Error.unprocessableEntity('Please provide a valid username or password'),
+      );
+      return;
+    }
+    const user = await User.findOne({ email: req.body.email });
+    if (!user) {
+      next(
+        Error.notFound('User is not signed up'),
+      );
+      return;
+    }
+    const validPassword = await bcrypt.compare(req.body.password, user.password);
```

Comment:
why not call this also isPasswordCorrect like line 95?

Reply:
because this checks if the user email is valid or not based on the login validation criteria , the isPasswordCorrect compares the user old and new password in the change passwrod function